text/3389-manifest-lint.md

```diff
@@ -0,0 +1,250 @@
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+A new `lints` table would be added to configure lints:
+```toml
+[lints]
+unsafe = "forbid"
+```
+and `cargo` would pass these along as flags to `rustc` and `clippy`.
+
+This would work with
+[RFC 2906 `workspace-deduplicate`](https://rust-lang.github.io/rfcs/2906-cargo-workspace-deduplicate.html?highlight=2906#):
+```toml
+[lints]
+workspace = true
+
+[workspace.lints]
+unsafe = "forbid"
+```
+
+## Documentation Updates
+
+## The `lints` section
+
+*as a new ["Manifest Format" entry](https://doc.rust-lang.org/cargo/reference/manifest.html#the-manifest-format)*
+
+Override the default level of lints by assigning them to a new level in a
+table, for example:
+```toml
+[lints]
+unsafe = "forbid"
+```
+
+Supported levels include:
+- `forbid`
+- `deny`
+- `warn`
+- `allow`
+
+**Note:** TOML does not support `:` in unquoted keys, requiring tool-specific
+lints to be quoted, like
+```toml
+[lints]
+"clippy::enum_glob_use" = "warn"
+```
+
+## The `lints` table
+
+*as a new [`[workspace]` entry](https://doc.rust-lang.org/cargo/reference/workspaces.html#the-workspace-section)*
+
+The `workspace.lints` table is where you define lint configuration to be inherited by members of a workspace.
+
+Specifying a workspace lint configuration is similar to package lints.
+
+Example:
+
+```toml
+# [PROJECT_DIR]/Cargo.toml
+[workspace]
+members = ["crates/*"]
+
+[workspace.lints]
+unsafe = "forbid"
+```
+
+```toml
+# [PROJECT_DIR]/crates/bar/Cargo.toml
+[package]
+name = "bar"
+version = "0.1.0"
+
+[lints]
+workspace = true
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+When parsing a manifest, cargo will resolve workspace inheritance for
+`lints.workspace = true` as it does with other fields.
+
+When running rustc, cargo will transform the lints from `lint = level` to
+`--level lint` and pass them on the command line before `RUSTFLAGS`, allowing
+user configuration to override package configuration.  These flags will be
+finterprinted so changing them will cause a rebuild.
+
+**Note:** This reserves the lint name `workspace` to allow workspace inheritance.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+A concern brought up in
+[rust-lang/rust-clippy#1313](https://github.com/rust-lang/rust-clippy/issues/1313)
+was that this will pass lints unconditionally to the underlying tool, leading
+to "undefined lint" warnings when used on earlier versions, requiring that
+warning to also be suppressed, reducing its value.  However, in the "Future
+possibility's", we mention direct support for tying lints to rust versions.
+
+This does not allow sharing lints across workspaces.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+This could be left to `clippy.toml` but that leaves `rustc` without a solution.
```

Comment:
The relationship between `clippy.toml` and `[lints]` table is not really clear here. Why do we need a new `[lints]` table given that we have `clippy.toml` today? Does `clippy.toml` become obsolete after this landed? How do we teach users to migrate one to the other, or the difference between these two files?

Sorry I don't have a constructive thought at this moment. Just try dumping things from my head.

Reply:
It is not possible to set lint levels in `clippy.toml`, which is a big motivation for this RFC. We don't want to add this to `clippy.toml`, because it is way better placed in `cargo`. If you want all the details about this topic, see this issue: https://github.com/rust-lang/rust-clippy/issues/1313

Reply:
The main deficiency with `clippy.toml` today is that you can't override the level for a lint.

The main deficiency in the long term for `clippy.toml` for this is that it is clippy specific but we need to also deal with rustc lints and rustdoc lints now and cargo and cargo-semver-check lints in the future.

An unknown for me for `clippy.toml` is its workspace interactions.

That said, I think this is a great conversation to have with the clippy team on what their ideal future state is and which better fits that ideal future state.

Reply:
From the Clippy side: We always treated `clippy.toml` as an unstable feature with the potential to be removed in the future.

With this RFC and the possibility to configure lints in `Cargo.toml` there is no more reason for `clippy.toml` to exist and we can deprecate/remove it.

So in a perfect world, lint level definition and lint configuration will be handled by `cargo` and `clippy.toml` doesn't exist.

Reply:
@flip1995 if this is accepted, what does the clippy team see as the role of `clippy.toml` in the long term?   If it is to be merged in to `Cargo.toml`, are there any deficiencies in this design for that, like clippy-wide configuration?  

Are there any challenges in reading configuration from `Cargo.toml` that would would need to resolve?  For example, would we need to include it in `cargo metadata` so we can process workspace inheritance for you?

Reply:
Thanks for those pointers and information! I feel like a summary of rust-lang/rust-clippy#1313 is worth mentioning somewhere in this RFC, if it doesn't burden the RFC and its author.

Reply:
`clippy.toml` allows lint configuration independent from lints. This is important because some configurations (most importantly `msrv`) is used by multiple lints. So in order to fully include the `clippy.toml` functionality in `Cargo.toml`, we'd need a way to specify keys that are **not** lints and then pass them to Clippy (or other tools).

We use `cargo metadata` (or better the cargo-metadata crate) to get information about the `Cargo.toml` file. So it would have to be included in the cargo metadata.

But I would leave this open for future discussions. For lint configurations, I'd suggest to add a `[lints.configuration]`/`[lints.<tool>.configuration]` subsection

Reply:
> I feel like a summary of [rust-lang/rust-clippy#1313](https://github.com/rust-lang/rust-clippy/issues/1313) is worth mentioning somewhere in this RFC

The summary is that `clippy.toml` can't define lint levels and will not do so in the future. The rest of the issue is patiently explaining this decision to people who kept asking (which was and is a reasonable question).

Reply:
> So in order to fully include the clippy.toml functionality in Cargo.toml, 

Is this something the RFC should aim for though?  I want to make sure this includes the clippy team's perspective rather than being prescriptive on what clippy should do.

> we'd need a way to specify keys that are not lints and then pass them to Clippy (or other tools).

Does cargo need to pass this extra configuration or just expose it via `cargo metadata`?

If we do need to pass it, I would suggest we do it as a literal translation of command-line arguments so cargo doesn't have to hard code how to pass each one.

> But I would leave this open for future discussions. For lint configurations, I'd suggest to add a [lints.configuration]/[lints.<tool>.configuration] subsection

A part of me wants to call this "metadata" for consistency (flags it as "cargo does not care") though I feel like that could also be confusing for new users not steeped in cargo.

This also has me leaning towards `lints.<tool>.<lint>` rather than `lints.<tool>::<lint>`.

Reply:
> No, this is too much for this RFC IMO. But we should keep this in mind for future possibilities.

Sorry, this is what I meant.  Should the RFC be written with the expectation that clippy would want to use lint configuration in `Cargo.toml`, replacing the use of `clippy.toml`.

Reply:
The one question I have with `lints.<tool>.<lint>`  is whether rustc should be under
- `lints`: cleaner translation for diagnostic messages to `Cargo.toml`
- `lints.rust`: cleaner data model, more future proofed against adding new tools.

Reply:
> Sorry, this is what I meant. Should the RFC be written with the expectation that clippy would want to use lint configuration in `Cargo.toml`, replacing the use of `clippy.toml`.

I would not put the main focus on this. I would focus on making defining lint levels most ergonomic. Only if it comes to 50:50 decisions I would go for the one that makes it easier to also include lint/tool configuration.

---

I would put `rustc` lints under just `[lints]`. In code Rust lints are also not scoped. So not scoping it in the Cargo.toml makes the most sense to me.

Reply:
I think we should stick with the top-level table, rather than making the default case more verbose for the sake of allowing for tool lints with the same prefix as an existing rust lint. I don't think that that case would work well with the compiler either.

Reply:
*https://github.com/rust-lang/rfcs/pull/3389#discussion_r1107371674 is the thread about `<level> = [<lint>]` or `<lint> = <level>`*

Thinking more about this, I don't think an unscoped rust lint table is feasible with `<lint> = <level>` as we need to scrape up all of the contents and then forward them on to a tool.   Without configurable lints, this is easy as we just say "is the value a string?  Forward it" but becomes more complex with configurable lints as now we are getting into detecting the shape of a table.

That said, I do not think this immediately pushes us from one data model to the other but this is another angle for us to consider.

Reply:
The RFC has been updated to `lints.<tool>.<level>` with a `lints.<tool>.metadata` table in the configurable lints future possibility

Reply:
I think this thread can be marked as resolved? All other topics opened up in this thread, like `<level> = [<lint>]` vs `<lint> = <level>` are addressed in different threads.